mlir/include/mlir/IR/CommonAttrConstraints.td

```diff
@@ -343,8 +343,10 @@ def SymbolNameAttr : StringBasedAttr<CPred<"::llvm::isa<::mlir::StringAttr>($_se
 
 // String attribute that has a specific value type.
 class TypedStrAttr<Type ty>
-    : StringBasedAttr<CPred<"::llvm::isa<::mlir::StringAttr>($_self)">,
-                            "string attribute"> {
+    : StringBasedAttr<
+        SubstLeaves<"$_self", "::mlir::cast<StringAttr>($_self).getType()",
```

Comment:
Can we assume at this point that the cast will succeed or should I test that `$_self` is `StringAttr` as well?

Reply:
I suspect that the order of the checks here matters. I.e., it should be safe to assume that it is a string attr. But to be sure, you could try to add another test case for `test.string_attr_with_type` in generic op syntax, where the attr is not a string attr.

Reply:
I had, since opening the PR, encountered a crash with the first version, which didn't test for `StringAttr`. The second commit fixed that but didn't add a test. A just pushed a third commit that adds a test that crashes without the fix from the second commit (but passes with the fix). Thanks for suggesting that!